README.md
# CYWAD

A tool automating gathering various numeric values as some kind of metrics from web pages )

[Live demo](https://cywad.herokuapp.com/) with [sources](https://github.com/estin/cywad-demo)

| WARNING: This is a hobby project to learn Rust! It works, but you can implement it in simpler manner, for example, by nodejs + puppeteer |
| --- |

Main features:
 - library or binary modes
 - `cli` mode
 - `server` mode to expose result via OpenAPI 3
 - configurable via TOML files
 - cron-like schedule
 - retry on errors
 - screenshots
 - server-sent events
 - PNG widget for embedding purposes
 - basic authorization using a header or by `?token=` query string parameter

Currently supports next `engines`:

 - embedded browser by [gtk-rs/webkit2gtk-rs](https://github.com/gtk-rs/webkit2gtk-rs)
 - external browser through chrome [devtools](https://chromedevtools.github.io/devtools-protocol/)


## Setup

Currently from sources only.

```bash
$ git clone https://github.com/estin/cywad.git
$ cd cywad
$ cargo build --release --features devtools,server,png_widget
```


## Example - Rust lang stars on github

Create `misc/github-rust.toml` with next content:

```toml

# target url
url = "https://github.com/rust-lang/rust"
# title
name = "rust github stars"

# cron like schedule for server mode
# https://crates.io/crates/cron
cron = "0   *   *     *       *  *  *"

# browser window size
window_width = 1280
window_height = 1024

# timeout to execute one step in ms
step_timeout = 15000

# timeout to wait befor start step in ms
step_interval = 1000

# retry in seconds
retry = [ 60, 300, 600 ]

# Step kind is one of
# - `wait` execute step until result would be `true`
# - `exec` execute code and go to the next step
# - `screenshot` take screenshot
# - `value` get numeric value

# Step0. wait before page is ready
[[steps]]
kind = "wait"
exec = """document.querySelector("a.js-social-count") ? true : false"""

# Step1. get count of stars and paint red border on stars badge
[[steps]]
kind = "value"
key = "stars"
exec = """(() => {
    const el = document.querySelector("a.js-social-count");

    // paint red border
    el.style.borderColor = "red";
    el.style.borderStyle = "solid";

    return parseInt(el.getAttribute("aria-label").split(' ')[0]);
})()"""

    # Determine level of value - any string
    [[steps.levels]]
    name = "green"
    more = 50000
 
    [[steps.levels]]
    name = "red"
    less = 30000 

    [[steps.levels]]
    name = "yellow"
    less = 50000

# Step2. take screenshot
[[steps]]
kind = "screenshot"
key = "stars"
```


## Example of `cli` mode:

Start Chrome devtools with command
```bash
$ chromium --remote-debugging-address=0.0.0.0 --remote-debugging-port=9222 --headless --disable-gpu --disable-software-rasterizer --disable-dev-shm-usage --no-sandbox --enable-logging --allow-running-insecure-content --ignore-certificate-errors
```

```bash
$ cargo run --release --features devtools -- cli -c misc/github-rust.toml
```

Result is
```json
[
  {
    "name": "rust github stars",
    "slug": "rust-github-stars",
    "datetime": "2019-12-21T23:47:34.710012303+03:00",
    "scheduled": null,
    "values": [
      {
        "key": "stars",
        "value": 41332.0,
        "level": "yellow"
      }
    ],
    "error": null,
    "screenshots": [
      {
        "name": "stars",
        "uri": "data:image/png;base64,<...skip...>"
      }
    ],
    "state": "Done",
    "previous": null,
    "steps_done": 3,
    "steps_total": 3,
    "attempt_count": 1
  }
]
```


## Example of `server` mode:


Start server
```bash
$ cargo run --release --features devtools,server -- serve -p misc --listen 127.0.0.1:8000
```

Get info
```bash
$ curl -s http://127.0.0.1:8000/api/items | jq
{
  "info": {
    "name": "cywad",
    "version": "0.1.0",
    "description": "",
    "server_datetime": "2019-12-21T23:52:32.331273021+03:00"
  },
  "items": [
    {
      "name": "rust github stars",
      "slug": "rust-github-stars",
      "datetime": "2019-12-21T23:51:48.608722134+03:00",
      "scheduled": "2019-12-22T08:00:00+03:00",
      "values": [
        {
          "key": "stars",
          "value": 41332,
          "level": "yellow"
        }
      ],
      "error": null,
      "screenshots": [
        {
          "name": "stars",
          "uri": "rust-github-stars/stars.png"
        }
      ],
      "state": "Done",
      "previous": null,
      "steps_done": 3,
      "steps_total": 3,
      "attempt_count": 1
    }
  ]
},
```

Get the screenshot
```bash
$ curl -v http://127.0.0.1:8000/screenshot/rust-github-stars/stars.png 2>&1 | grep "<"
< HTTP/1.1 200 OK
< content-length: 223039
< content-type: image/png
Binary file (standard input) matches
```

You can check out [openapi.yaml](openapi.yaml) for all available rest API methods.

## Testing `devtools` engine

```bash
$ chromium -remote-debugging-address=0.0.0.0 --remote-debugging-port=9222 --headless --disable-gpu --disable-software-rasterizer --disable-dev-shm-usage --no-sandbox --enable-logging --allow
-running-insecure-content --ignore-certificate-errors
```

```bash
$ cargo test --features=devtools,server,png_widget -- --nocapture
```

## Testing `webkit2gtk` engine

```bash
$ cargo test --features=webkit,server,png_widget -- --nocapture
``` 

## License

This project is licensed under the [MIT license](LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed
- split to crates, refactor
- refactor: ws client actor logic 

### Fixed
- mark result as failed on general engine error to prevent hangs
- scheduler logic
- update result item datetime on start processing 
- ws request timeout
- don't hang on worth devtools config response 

### Added
- frontend on yew introduced

## [0.1.2] - 2020-08-16

### Fixed
- webkit tests

### Changed
- webkit: update gtk to 0.9.1
- use serde with derive feature instead of serde_derive, fmt&clippy 
- prevent panic if step index out of bounds
- don't send Network.clearBrowserCache and Network.clearBrowserCookies

## [0.1.1] - 2020-01-14

### Fixed
- update `scheduled` attribute on any item state

### Changed
- Edition 2018: wip
- up to date actix/tokio/futures/bytes and others
- http server starts default workers count
- clippy and rustfmt fixes

## [0.1.0] - 2020-01-02

### Added

- First public release


cywad-core/src/lib.rs
use chrono::prelude::{DateTime, Local};
use log::{debug, error};
use serde::{Deserialize, Serialize};
use std::fmt;
use std::fs::File;
use std::io::Read;
use std::sync::mpsc::Sender;
use std::sync::{Arc, Mutex, RwLock};

#[cfg(feature = "server")]
use std::str::FromStr;

#[cfg(feature = "server")]
use cron::Schedule;

use anyhow::Error;
use anyhow::{anyhow, bail};
use slug::slugify;

pub const APP_NAME: Option<&'static str> = option_env!("CARGO_PKG_NAME");
pub const APP_DESCRIPTION: Option<&'static str> = option_env!("CARGO_PKG_DESCRIPTION");
pub const APP_VERSION: Option<&'static str> = option_env!("CARGO_PKG_VERSION");
pub const SCHEDULER_SLEEP: u64 = 60; // each minute

#[derive(Debug, Serialize, Deserialize)]
pub struct AppInfo {
    pub name: String,
    pub version: String,
    pub description: String,
    pub server_datetime: DateTime<Local>,
}

impl AppInfo {
    pub fn default() -> Self {
        AppInfo {
            name: APP_NAME.unwrap_or("n/a").to_string(),
            description: APP_DESCRIPTION.unwrap_or("n/a").to_string(),
            version: APP_VERSION.unwrap_or("n/a").to_string(),
            server_datetime: Local::now(),
        }
    }
}

#[derive(Debug, Deserialize, Clone)]
pub struct Config {
    pub url: String,
    pub name: String,
    pub cron: Option<String>,
    pub retry: Option<Vec<i64>>,
    pub window_width: u32,
    pub window_height: u32,
    pub step_timeout: i64,
    pub step_interval: u32,
    pub steps: Vec<StepConfig>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct StepConfig {
    pub kind: StepKind,
    pub key: Option<String>,
    pub exec: Option<String>,
    pub levels: Option<Vec<LevelConfig>>,
}

#[derive(Debug, Deserialize, PartialEq, Clone)]
#[serde(rename_all = "lowercase")]
pub enum StepKind {
    Wait,
    Exec,
    Value,
    Screenshot,
}

#[derive(Debug, PartialEq, Clone)]
pub enum StepResult {
    Idle,
    InWork,
    Repeat,
    Next,
    Error,
}

#[derive(Debug, Clone)]
pub struct ExecutionContext {
    pub ts_start: i64,
    pub step_result: StepResult,
    pub step_index: usize,
    pub step_max_index: usize,
    pub config: Config,
    pub config_index: usize,
    pub state: SharedState,
}

#[derive(Debug, Clone, Deserialize)]
pub struct LevelConfig {
    pub name: String,
    pub more: Option<f64>,
    pub less: Option<f64>,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub enum ResultItemState {
    Idle,
    InQueue,
    InWork,
    Done,
    Err,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ValueItem {
    pub key: String,
    pub value: f64,
    pub level: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ScreenshotItem {
    pub name: String,
    pub uri: String,

    #[serde(skip_serializing)]
    #[serde(skip_deserializing)]
    pub data: Vec<u8>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct PreviouResultItemState {
    pub datetime: DateTime<Local>,
    pub values: Vec<ValueItem>,
    pub screenshots: Vec<ScreenshotItem>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ResultItem {
    pub name: String,
    pub slug: String,
    pub datetime: DateTime<Local>,
    pub scheduled: Option<DateTime<Local>>,
    pub values: Vec<ValueItem>,
    pub error: Option<String>,
    pub screenshots: Vec<ScreenshotItem>,
    pub state: ResultItemState,
    pub previous: Option<PreviouResultItemState>,
    pub steps_done: Option<usize>,
    pub steps_total: Option<usize>,
    pub attempt_count: Option<usize>,
}

impl ResultItem {
    pub fn new(name: &str) -> Self {
        ResultItem {
            name: name.to_string(),
            slug: slugify(name.to_string()),
            datetime: Local::now(),
            scheduled: None,
            values: Vec::new(),
            error: None,
            screenshots: Vec::new(),
            state: ResultItemState::Idle,
            previous: None,
            steps_done: None,
            steps_total: None,
            attempt_count: None,
        }
    }

    pub fn is_ok(&self) -> bool {
        self.state == ResultItemState::Done
    }
    pub fn is_err(&self) -> bool {
        self.state == ResultItemState::Err
    }
    pub fn is_in_work(&self) -> bool {
        self.state == ResultItemState::InWork
    }
    pub fn is_idle(&self) -> bool {
        self.state == ResultItemState::Idle
    }
    pub fn is_in_queue(&self) -> bool {
        self.state == ResultItemState::InQueue
    }
    pub fn get_error(&self) -> &str {
        self.error.as_ref().map_or("None", |error| error)
    }
    pub fn get_scheduled(&self) -> String {
        self.scheduled
            .as_ref()
            .map_or("-".to_string(), |scheduled| {
                scheduled.format("%Y-%m-%d %H:%M:%S").to_string()
            })
    }
}

pub fn get_level(value: f64, levels: &[LevelConfig]) -> Option<String> {
    for level in levels {
        if let Some(more_value) = level.more {
            if value >= more_value {
                return Some(level.name.to_owned());
            }
        }
        if let Some(less_value) = level.less {
            if value < less_value {
                return Some(level.name.to_owned());
            }
        }
    }
    None
}

pub fn validate_config(config: &Config) -> Result<(), Error> {
    // validate schedule
    #[cfg(feature = "server")]
    {
        if let Some(ref schedule) = config.cron {
            if let Err(e) = Schedule::from_str(schedule) {
                return Err(anyhow!("'cron' field - {}", e));
            }
        }
    }

    // validate each step
    for (i, step) in config.steps.iter().enumerate() {
        match step.kind {
            StepKind::Wait | StepKind::Value | StepKind::Exec => {
                // `exec` field required for wait/value/exec
                if step.exec.is_none() {
                    return Err(anyhow!(
                        "'wait/value/exec' step #{} without 'exec' field",
                        i + 1
                    ));
                }
                // `value` step must be with `key` field
                if step.kind == StepKind::Value && step.key.is_none() {
                    return Err(anyhow!("'value' step #{} without 'key' field", i + 1));
                }
            }
            _ => {}
        }
    }
    Ok(())
}

pub fn load_config(path: &str) -> Result<Config, Error> {
    let mut file = File::open(path)?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    let config: Config =
        toml::from_str(&contents).map_err(|e| anyhow!("Toml file parsing error: {}", e))?;
    validate_config(&config)?;
    Ok(config)
}

#[derive(Debug)]
pub struct State {
    pub configs: Vec<Config>,
    pub results: Vec<ResultItem>,
    pub tx: Option<Mutex<Sender<usize>>>,
    pub tx_vec: Option<Vec<Mutex<Sender<ResultItem>>>>,
}

impl State {
    pub fn broadcast(&mut self, index: usize) {
        self.tx_vec = {
            let tx_vec = match self.tx_vec {
                Some(ref mut vec) => vec,
                None => return,
            };
            let mut new_tx_vec: Vec<Mutex<Sender<ResultItem>>> = Vec::new();
            if let Some(ref mut item) = self.results.get(index) {
                // TODO change to drain filter
                for tx in tx_vec.drain(..) {
                    let success: bool = {
                        if let Ok(sender) = tx.lock() {
                            sender.send(item.clone()).is_ok()
                        } else {
                            error!("Sender lock error");
                            false
                        }
                    };
                    if success {
                        new_tx_vec.push(tx);
                    }
                }
            }
            Some(new_tx_vec)
        }
    }

    pub fn mark_as_err(&mut self, index: usize) {
        debug!("Result item #{} mark as failed", index);
        if let Some(item) = self.results.get_mut(index) {
            item.state = ResultItemState::Err;
            item.datetime = Local::now();
            item.values.clear();
            item.screenshots.clear();
            item.steps_done = None;
            item.steps_total = None;
            item.attempt_count = Some(item.attempt_count.unwrap_or(0) + 1);
        }
        self.broadcast(index);
    }
}

pub type SharedState = Arc<RwLock<State>>;

impl fmt::Display for ExecutionContext {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{} {}/{}",
            self.config.name, self.step_index, self.step_max_index,
        )
    }
}

impl ExecutionContext {
    pub fn new(config: Config, config_index: usize, state: SharedState) -> Self {
        ExecutionContext {
            ts_start: Local::now().timestamp(),
            step_index: 0,
            step_result: StepResult::Idle,
            step_max_index: config.steps.len() - 1,
            config,
            config_index,
            state,
        }
    }

    pub fn is_done(&self) -> bool {
        self.step_index > self.step_max_index
    }

    pub fn get_step(&self) -> Result<StepConfig, Error> {
        if self.is_done() {
            bail!("[{}] Step #{} out of bounds", self, self.step_index);
        }
        Ok(self
            .config
            .steps
            .get(self.step_index)
            .ok_or_else(|| anyhow!("Step not found by index {}", self.step_index))?
            .clone())
    }

    pub fn start_new_step(&mut self) -> Result<(), Error> {
        // increment step indexes
        self.step_index += 1;
        // start new timer
        self.ts_start = Local::now().timestamp();
        match self.state.write() {
            Ok(mut state) => {
                if let Some(item) = state.results.get_mut(self.config_index) {
                    item.datetime = Local::now();
                    item.steps_done = Some(self.step_index);
                }
                state.broadcast(self.config_index);
            }
            Err(e) => return Err(anyhow!("RWLock error: {}", e)),
        };
        debug!("[{}] Step index increased #{}", self, self.step_index);
        Ok(())
    }

    pub fn timeout(&mut self) -> Result<(), Error> {
        match self.state.write() {
            Ok(mut state) => {
                if let Some(item) = state.results.get_mut(self.config_index) {
                    item.datetime = Local::now();
                    item.state = ResultItemState::Err;
                    item.error = Some(format!("Timeout: {}", self.config.step_timeout));
                }
                state.broadcast(self.config_index);
            }
            Err(e) => return Err(anyhow!("RWLock error: {}", e)),
        };
        Ok(())
    }

    pub fn done(&mut self) -> Result<(), Error> {
        match self.state.write() {
            Ok(mut state) => {
                if let Some(item) = state.results.get_mut(self.config_index) {
                    item.datetime = Local::now();
                    item.state = ResultItemState::Done;
                    item.steps_done = item.steps_total;
                }
                state.broadcast(self.config_index);
            }
            Err(e) => return Err(anyhow!("RWLock error: {}", e)),
        };
        Ok(())
    }

    pub fn error(&mut self, error_message: String) -> Result<(), Error> {
        match self.state.write() {
            Ok(mut state) => {
                if let Some(item) = state.results.get_mut(self.config_index) {
                    item.datetime = Local::now();
                    item.state = ResultItemState::Err;
                    item.error = Some(error_message);
                }
                state.broadcast(self.config_index);
            }
            Err(e) => return Err(anyhow!("RWLock error: {}", e)),
        };
        Ok(())
    }

    pub fn add_value_and_start_new_step(&mut self, value: f64) -> Result<(), Error> {
        match self.state.write() {
            Ok(mut state) => {
                if let Some(item) = state.results.get_mut(self.config_index) {
                    let step = &self.get_step()?;

                    let key = if let Some(ref key) = step.key {
                        key.to_owned()
                    } else {
                        if step.kind == StepKind::Value {
                            unreachable!("'key' not defined for step value");
                        }
                        "".to_owned()
                    };

                    item.values.push(ValueItem {
                        value,
                        key,
                        level: if let Some(ref levels) = step.levels {
                            get_level(value, levels)
                        } else {
                            None
                        },
                    });

                    item.datetime = Local::now();
                    self.step_index += 1;
                    self.ts_start = Local::now().timestamp();
                    item.steps_done = Some(self.step_index);
                }
                state.broadcast(self.config_index);
            }
            Err(e) => return Err(anyhow!("RWLock error: {}", e)),
        }

        Ok(())
    }

    pub fn add_screenshot_and_start_new_step(&mut self, value: Vec<u8>) -> Result<(), Error> {
        match self.state.write() {
            Ok(mut state) => {
                if let Some(item) = state.results.get_mut(self.config_index) {
                    let step = &self.get_step()?;
                    let (name, for_uri) = if let Some(ref key) = step.key {
                        (key.to_owned(), key.to_owned())
                    } else {
                        ("".to_string(), item.screenshots.len().to_string())
                    };

                    item.screenshots.push(ScreenshotItem {
                        uri: format!("{}/{}.png", item.slug, slugify(&for_uri)),
                        name,
                        data: value,
                    });

                    item.datetime = Local::now();
                    self.step_index += 1;
                    self.ts_start = Local::now().timestamp();
                    item.steps_done = Some(self.step_index);
                }
                state.broadcast(self.config_index);
            }
            Err(e) => return Err(anyhow!("RWLock error: {}", e)),
        }

        Ok(())
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct HeartBeat {
    pub server_datetime: DateTime<Local>,
}

impl HeartBeat {
    pub fn default() -> Self {
        HeartBeat {
            server_datetime: Local::now(),
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ItemsResponse {
    pub info: AppInfo,
    pub items: Vec<ResultItem>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ItemResponse {
    pub server_datetime: DateTime<Local>,
    pub item: ResultItem,
}

#[derive(Clone)]
#[cfg(feature = "webkit")]
pub struct EngineOptions {
    pub keep_open: bool,
}

#[derive(Clone)]
#[cfg(feature = "devtools")]
pub struct EngineOptions {
    pub command: Option<String>,
    pub endpoint: String,
    pub http_timeout: u64,
    pub max_frame_size: usize,
}

#[cfg(any(feature = "devtools", feature = "webkit"))]
impl Default for EngineOptions {
    #[cfg(feature = "webkit")]
    fn default() -> Self {
        EngineOptions { keep_open: false }
    }

    #[cfg(feature = "devtools")]
    fn default() -> Self {
        EngineOptions {
            command: None,
            endpoint: "http://127.0.0.1:9222/json".into(),
            http_timeout: 5,
            max_frame_size: 3 * 1024 * 1024,
        }
    }
}

#[cfg(any(feature = "devtools", feature = "webkit"))]
pub trait EngineTrait {
    fn execute(
        &mut self,
        index: usize,
        state: SharedState,
        engine_options: EngineOptions,
    ) -> Result<(), Error>;
}
